File: src/Day3RucksackReorg/c/main.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <ctype.h>
#include <direct.h>

#define MAX_LINE_LENGTH 100

char get_common_char_from_split_line(const char *s)
{
    int len = strlen(s);
    int half_len = len / 2;

    for (int i = 0; i < half_len; i++)
    {
        for (int j = half_len; j < len; j++)
        {
            if (s[i] == s[j])
            {
                return s[i];
            }
        }
    }

    return '\0';
}

int get_alphabetical_index_of_char(char c)
{
    return c % 32;
}

char find_common_char_across_lines(char *lines[], int num_lines)
{
    for (int i = 0; lines[0][i] != '\0'; i++)
    {
        char c = lines[0][i];
        int found = 1;

        for (int j = 1; j < num_lines; j++)
        {
            if (strchr(lines[j], c) == NULL)
            {
                found = 0;
                break;
            }
        }

        if (found)
        {
            return c;
        }
    }

    return '\0';
}

int main()
{
    // Get input file path
    char filePath[256] = "";
    sprintf(filePath, "%s\\input.txt", _getcwd(NULL, 0));
    FILE *fpointer = fopen(filePath, "r"); // r file mode to read contents of the files
    char *line = NULL;
    size_t lineLength = 0;

    int challenge_1_total_value = 0;
    int challenge_2_total_value = 0;
    int challenge_2_chunk_length = 3;
    int challenge_2_index = 1;
    char *challenge_2_line_chunk[3];

    while (getline(&line, &lineLength, fpointer) != -1)
    {
        char common_character = get_common_char_from_split_line(line);
        int char_index = get_alphabetical_index_of_char(common_character);

        if (isupper(common_character))
        {
            challenge_1_total_value += 26;
        }

        challenge_1_total_value += char_index;

        if (challenge_2_index < challenge_2_chunk_length)
        {
            challenge_2_line_chunk[challenge_2_index - 1] = malloc(MAX_LINE_LENGTH * sizeof(char));
            strcpy(challenge_2_line_chunk[challenge_2_index - 1], line);
            challenge_2_index++;
        }
        else
        {
            challenge_2_line_chunk[challenge_2_index - 1] = malloc(MAX_LINE_LENGTH * sizeof(char));
            strcpy(challenge_2_line_chunk[challenge_2_index - 1], line);

            char common_character2 = find_common_char_across_lines(challenge_2_line_chunk, challenge_2_chunk_length);
            int char_index2 = get_alphabetical_index_of_char(common_character2);

            if (isupper(common_character2))
            {
                challenge_2_total_value += 26;
            }

            challenge_2_total_value += char_index2;
            challenge_2_index = 1;
        }
    }

    for (int i = 0; i < challenge_2_chunk_length; i++)
    {
        free(challenge_2_line_chunk[i]);
    }

    fclose(fpointer);

    printf("Challenge 1: %i\n", challenge_1_total_value);
    printf("Challenge 2: %i\n", challenge_2_total_value);

    return 0;
}

File: src/Day3RucksackReorg/rs/src/main.rs
use std::env;
use std::fs::File;
use std::io::{BufRead, BufReader};

fn get_common_char_from_split_line(s: &str) -> Option<char> {
    let len = s.len();
    let half_len = len / 2;

    for (_i, c1) in s.chars().take(half_len).enumerate() {
        for c2 in s.chars().skip(half_len) {
            if c1 == c2 {
                return Some(c1);
            }
        }
    }

    None
}

fn get_alphabetical_index_of_char(c: char) -> usize {
    c as usize % 32
}

fn find_common_char_across_lines(lines: &[String]) -> Option<char> {
    for (_i, c) in lines[0].chars().enumerate() {
        let mut found = true;

        for line in &lines[1..] {
            if !line.contains(c) {
                found = false;
                break;
            }
        }

        if found {
            return Some(c);
        }
    }

    None
}

fn main() {
    let input_path = env::current_dir()
        .unwrap()
        .parent()
        .unwrap()
        .join("input.txt");
    let file = File::open(input_path).unwrap();
    let reader = BufReader::new(file);
    let mut challenge_1_total_value = 0;
    let mut challenge_2_total_value = 0;
    let mut challenge_2_chunk: Vec<String> = Vec::new();

    for (i, line) in reader.lines().enumerate() {
        let line = line.unwrap();
        let common_character = get_common_char_from_split_line(&line);
        let char_index = common_character
            .map(get_alphabetical_index_of_char)
            .unwrap_or(0);

        if let Some(c) = common_character {
            if c.is_ascii_uppercase() {
                challenge_1_total_value += 26;
            }

            challenge_1_total_value += char_index;
        }

        if i % 3 == 2 {
            challenge_2_chunk.push(line);

            if let Some(common_character2) = find_common_char_across_lines(&challenge_2_chunk) {
                let char_index2 = get_alphabetical_index_of_char(common_character2);

                if common_character2.is_ascii_uppercase() {
                    challenge_2_total_value += 26;
                }

                challenge_2_total_value += char_index2;
            }

            challenge_2_chunk.clear();
        } else {
            challenge_2_chunk.push(line);
        }
    }

    println!("Challenge 1: {}", challenge_1_total_value);
    println!("Challenge 2: {}", challenge_2_total_value);
}
